### PR TITLE
Introduce todo list form

### PR DESCRIPTION
**Pull request by trofimovama**
Closes: {link to STORY_ID}

---

##### Extra Details:

- If you made something additional, not related to the initial issue scope, put it here in this section
- Otherwise, completely remove the section called `Extra Details` as it stated in the [gitflow](../docs/git-flow.md#оформление_pr)

### CODE REVIEW

**Review by vladlen-volkov — CHANGES_REQUESTED** [review]
Almost there!

File: src/index.js [hunk 1]
```diff
@@ -1,3 +1,32 @@
...
+ *             }
+ *           ]
+ *   },
+ *   ...
+ * ]
+ * }}
+ */
+
 import { renderApp } from './components/render-app.component';
+import { restoreState, rootState } from '../state/root.state';
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/index.js:29]: Moved

File: src/index.js [hunk 2]
```diff
@@ -1,3 +1,32 @@
...
 
-renderApp();
+renderApp(rootState);
+restoreState();
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/index.js:32]: Removed

File: src/components/render-app.component.js [hunk 3]
```diff
@@ -1,24 +1,33 @@
...
+    rootContainer.append(elem);
   });
 
-  rootContainer.addEventListener('click', (e) => {
-    if (!editableForm.contains(e.target)) {
-      formOverlay.style.display = 'none';
-    }
+  roundButton.addEventListener('click', () => {
+    const formDraftObject = { title: '', items: [] };
+    formState.push(formDraftObject);
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-app.component.js:20]: Removed

[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-app.component.js:20]: Fixed while moving card render form `renderOverlay` component to `renderApp` component. State is updated if no filled data

File: src/components/render-overlay.component.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
+/* eslint-disable import/no-cycle */
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-overlay.component.js:1]: Yes, it is. Removed

File: src/components/render-form.component.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
+/* eslint-disable import/no-cycle */
+/* eslint-disable array-callback-return */
+/* eslint-disable no-param-reassign */
+import { renderTemplate } from '../../utils/render-template';
+import { notifyStateUpdated, rootState } from '../../state/root.state';
+import { renderCard } from './render-card.component';
+
+import '../index.scss';
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-form.component.js:8]: Done

File: src/components/render-form.component.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  const cardContainer = editableForm.querySelector('.card-form');
+  const formTitle = editableForm.querySelector('.card-form__title');
+  const addFormItemIcon = editableForm.querySelector('.add-item__icon');
+  const addFormItemInput = editableForm.querySelector('.add-item__input');
+  const formListContainer = editableForm.querySelector('.card-form__list');
+  const formDeleteIcon = editableForm.querySelector('.card-footer__icon');
+
+  formTitle.textContent = formState.title;
+
+  formState.items.map((el) => {
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-form.component.js:43]: Extracted

File: src/components/render-form.component.js [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+  formTitle.addEventListener(
+    'change',
+    () => {
+      formState.title = formTitle.value;
+    },
+    { once: true },
+  );
+
+  const clickItem = () => {
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-form.component.js:110]: Renamed

File: src/components/render-form.component.js [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+      formListItemRemoveIcon.addEventListener('click', () => {
+        formListItem.remove();
+        formState.items.splice(i, 1);
+      });
+    }
+    addFormItemInput.value = '';
+  };
+
+  const enterItem = (e) => {
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-form.component.js:165]: Renamed

File: src/components/render-form.component.js [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    if (e.key === 'Enter') {
+      clickItem();
+    }
+  };
+
+  addFormItemIcon.addEventListener('click', clickItem);
+  addFormItemInput.addEventListener('keypress', enterItem);
+
+  formDeleteIcon.addEventListener(
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-form.component.js:174]: Done

File: src/components/render-card.component.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
+/* eslint-disable no-param-reassign */
+/* eslint-disable import/no-cycle */
+/* eslint-disable array-callback-return */
+import { notifyStateUpdated, rootState } from '../../state/root.state';
+import { renderTemplate } from '../../utils/render-template';
+import { rootContainer, renderForm } from './render-form.component';
+import { renderOverlay } from './render-overlay.component';
+import '../index.scss';
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-card.component.js:8]: Removed

File: src/components/render-card.component.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+      notifyStateUpdated();
+    },
+    { once: true },
+  );
+
+  cardDeleteIcon.addEventListener(
+    'click',
+    () => {
+      cardState = null;
+      const i = rootState.indexOf(cardState);
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-card.component.js:105]: Done

File: src/components/render-form.component.js [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+        <input type="checkbox" class="checkbox__input" />
+        <span class="checkbox__decor">
+          <svg class="checkbox__decor-icon" fill="#fff" width="12" height="12">
+            <use xlink:href="#tick"></use>
+          </svg>
+        </span>
+      </label>
+      <div class="flexible-wrap">
+        <textarea
+          value = ${el.text}
```
[the remark this answers is not in the document]
Comment (vladlen-volkov) [reply src/components/render-form.component.js]: Meant to put in the quotes also 😣
Sorry for nitpicking, you don't have to fix it if you don't want bc it's a minor

File: src/components/render-list-item.component.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+  const formItemText = item.querySelector('.card-form__content');
+  const formCheckboxInput = item.querySelector('.checkbox__input');
+
+  formItemText.value = itemState.text;
+
+  if (itemState.isDone) {
+    formItemText.classList.add('list-item__content--crossed');
+    formCheckboxInput.setAttribute('checked', 'true');
+    itemState.isDone = true;
```
Comment (vladlen-volkov) [inline src/components/render-list-item.component.js:37]: Why do we need assignment here? (same with L:40)

File: src/components/render-form.component.js [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+  const cardContainer = editableForm.querySelector('.card-form');
+  const formTitle = editableForm.querySelector('.card-form__title');
+  const addFormItemIcon = editableForm.querySelector('.add-item__icon');
+  const addFormItemInput = editableForm.querySelector('.add-item__input');
+  const formListContainer = editableForm.querySelector('.card-form__list');
+  const formDeleteIcon = editableForm.querySelector('.card-footer__icon');
+
+  const itemState = formState.items;
+
+  itemState.map((el) => {
```
Comment (vladlen-volkov) [inline src/components/render-form.component.js:37]: What is the difference between `forEach` and `map`?

**Review by vladlen-volkov — CHANGES_REQUESTED** [review]
Way better, but it still needs refinements 😣

⚠️ And I'd like to introduce one more requirement: now you aren't allowed to import anything from `state` folder besides in the `render-app.component` 

C'mon! You can do this! 💪

File: src/components/render-app.component.js [hunk 15]
```diff
@@ -1,24 +1,57 @@
...
-  rootContainer.addEventListener('click', (e) => {
-    if (!editableForm.contains(e.target)) {
-      formOverlay.style.display = 'none';
-    }
+    const form = renderForm(formDraftObject, () => {
+      const i = formState.indexOf(formDraftObject);
+      formState.splice(i, 1);
+      notifyStateUpdated();
+    });
+    formState.push(formDraftObject);
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-app.component.js]: Done

File: src/components/render-app.component.js [hunk 16]
```diff
@@ -1,24 +1,57 @@
...
+
+    const overlay = renderOverlay(form.elem, () => {
+      if (formDraftObject.title.trim() !== '' || formDraftObject.items.length !== 0) {
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-app.component.js]: Well-received) Syntax improved

File: src/components/render-overlay.component.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
+import { renderTemplate } from '../utils/render-template';
+
+export const renderOverlay = (elem, onCardRender) => {
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-overlay.component.js]: Renamed

File: src/components/render-overlay.component.js [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+  const overlayTemplate = renderTemplate(`
+  <div class="overlay">
+      <div class="overlay-container"></div>
+    </div>
+`);
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-overlay.component.js]: Fixed

File: src/components/render-list-item.component.js [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+  };
+
+  formCheckboxInput.addEventListener('click', handleCheckboxUpdated);
+
+  formItemText.addEventListener(
+    'change',
+    () => {
+      itemState.text = formItemText.value;
+    },
+    { once: true },
```
[the remark this answers is not in the document]
Comment (trofimovama) [reply src/components/render-list-item.component.js]: Removed

File: src/components/render-list-item.component.js (hunk 13 above)
Comment (trofimovama) [reply src/components/render-list-item.component.js]: We do not need) Removed

File: src/components/render-form.component.js (hunk 14 above)
Comment (trofimovama) [reply src/components/render-form.component.js:36]: `map` returns new array according to result provided by the function we call. forEach does not. It returns undefined. They both mutate the original array (by means of callback provided)

**Review by vladlen-volkov — APPROVED** [review]

**Pull request merged**